File: backend/app/schemas/system_status.py
from pydantic import BaseModel
from typing import List, Dict, Any, Optional, Literal
from datetime import datetime


class ServiceStatus(BaseModel):
    name: str
    status: Literal["healthy", "degraded", "error", "unknown", "not_configured"]
    message: Optional[str] = None
    latencyMs: Optional[float] = None
    details: Optional[Dict[str, Any]] = None


class QueueStatus(BaseModel):
    name: str
    depth: int
    workerCount: int


class CeleryTaskStatus(BaseModel):
    taskName: str
    displayName: str
    schedule: str
    lastRunTime: Optional[datetime] = None
    nextRunTime: Optional[datetime] = None
    lastDurationMs: Optional[int] = None
    lastStatus: Optional[str] = None
    status: Literal["running", "idle", "failed", "unknown"]


class DownloadClientStatus(BaseModel):
    id: int
    name: str
    clientType: str
    status: Literal["connected", "error", "disabled"]
    version: Optional[str] = None
    message: Optional[str] = None


class ExternalApiStatus(BaseModel):
    name: str
    status: Literal["healthy", "error", "not_configured"]
    message: Optional[str] = None
    lastChecked: Optional[datetime] = None


class SystemStatusResponse(BaseModel):
    timestamp: datetime
    version: str
    database: ServiceStatus
    cache: ServiceStatus
    pgBouncer: ServiceStatus
    celery: ServiceStatus
    queues: List[QueueStatus]
    celeryTasks: List[CeleryTaskStatus]
    downloadClients: List[DownloadClientStatus]
    externalApis: Dict[str, ExternalApiStatus]

File: frontend/src/types/system-status.ts
export type ServiceStatusLevel =
	| "healthy"
	| "degraded"
	| "error"
	| "unknown"
	| "not_configured";
export type TaskStatusLevel = "running" | "idle" | "failed" | "unknown";
export type DownloadClientStatusLevel = "connected" | "error" | "disabled";
export type ApiStatusLevel = "healthy" | "error" | "not_configured";

export interface ServiceStatus {
	name: string;
	status: ServiceStatusLevel;
	message: string | null;
	latencyMs: number | null;
	details?: Record<string, unknown>;
}

export interface QueueStatus {
	name: string;
	depth: number;
	workerCount: number;
}

export interface CeleryTaskStatus {
	taskName: string;
	displayName: string;
	description: string;
	schedule: string;
	lastRunTime: string | null;
	nextRunTime: string | null;
	lastDurationMs: number | null;
	lastStatus: string | null;
	status: TaskStatusLevel;
}

export interface DownloadClientStatus {
	id: number;
	name: string;
	clientType: string;
	status: DownloadClientStatusLevel;
	version: string | null;
	message: string | null;
}

export interface ExternalApiStatus {
	name: string;
	status: ApiStatusLevel;
	message: string | null;
	lastChecked: string | null;
}

export interface SystemStatusResponse {
	timestamp: string;
	version: string;
	database: ServiceStatus;
	cache: ServiceStatus;
	pgBouncer: ServiceStatus;
	celery: ServiceStatus;
	queues: QueueStatus[];
	celeryTasks: CeleryTaskStatus[];
	downloadClients: DownloadClientStatus[];
	externalApis: Record<string, ExternalApiStatus>;
}
